[CAM] Add UI elements for viewing and editing tool controller parameters from within the operation edit UI

davidgilkaufman opened the pull request:
This PR is part of a [freecad grant](https://github.com/FreeCAD/FPA-grant-proposals/issues/46). It satisfies the first deliverable of sub-project 1. See the grant description for the motivation of this PR.
[Forum post](https://forum.freecad.org/viewtopic.php?t=99180)
Main issue: #23179 
Also fixes: #19077 

This PR adds UI elements for editing an operation's tool controller's values from within the operation's edit UI.

<img width="416" height="152" alt="20250818_11h28m15s_grim" src="https://github.com/user-attachments/assets/bd8ca0ed-71e4-4fef-97bd-11d27d3c1b95" />

<img width="401" height="545" alt="20250818_11h29m02s_grim" src="https://github.com/user-attachments/assets/f0153b49-c85b-4eaf-a3a1-5ca8c18dfa76" />

## Notable new UI elements/interactions

- Each CAM operation now has a checkbox labeled "Edit Tool Controller" below the combo box for selecting the operation's tool controller. When checked/unchecked it shows/hides the editor UI items. By default they are hidden, to keep the UI clean.
- Most UI items are exactly the same as in the standalone tool controller editor, and are controlled by (mostly) the same code to keep the different versions of the tool controller edit UI in sync
- In addition to these normal items, there is a new "Copy Tool Controller" button. When clicked, it creates a new tool controller with the same values as the current one, and uses that new controller for the operation. This button exists to make it easy for users to change the speeds and feeds for only the current operation (and not any other operations that use the current tool controller). The button is visible only if the tool controller is shared by other operations

In addition to the primary changes in this PR, I did a drive-by fix of the Probe operation. There was a bug likely introduced with the new tool library, where Probe referred to its tool's ShapeName, but that property was renamed (and updated slightly) to ShapeType, preventing the probe operation from working. I fixed this while testing my UI changes for the probe operation.


## Notes for review and testing

Reviewing the code changes for this PR should be fairly easy, since it's not that many lines changed, and many of them are UI changes. However, validating correct behavior will be more involved, since this change touches the UI panel of every operation (including exprerimental operations, which you need to have enabled to test!)

To test the functionality of this PR, I have done a robust test of the new UI elements on the Profile operation, and a more cursory test of the UI on every other operation. In the review process, I think it would be prudent for a second person to do this as well to ensure I haven't missed anything, failed to redo a check originally done on an old version, etc.

Update: Additional notes, for the "new tool controller..." function: I run Qt6 on Linux, and this needs to be tested on Qt5 (preferably non-Linux). I want to know that the new TC dialog workflow runs cleanly, including if the user opens the tool library from that dialog.

Connor9220 commented:
So,  The tool number is part of the tool, not the tool controller.  You define the tool number in the tool library.  The tool controller is responsible for the feeds and speeds.   We 100% don't want to allow someone to change the tool number in the operations panel.   If this is something someone needs, then it needs to under a preference toggle or something, because that's a good way to break a tool or machine.  

Tool number != Slot/Rack/Pocket. While they can be the same value, they're not the same.  In some CNC machines, the tool will actually end up in a different pockets each time the tool is changed.  Slot/Rack/Pocket isn't something the CAM software is aware of.  That should be handled with the tool controller.  What controller are you using?  I've never seen one not understand tool number vs slot/rack/pocket #.  That's really odd.

What I'm suggesting isn't really conversational.   The underlying objects can stay the same, I'm just suggesting some terminology changes, and adding a way to group the tool controllers related to chosen tool.  This should all be GUI related.

tarman3 commented:
> My other thought, is this: I don't think we need to use the term Tool Controller. This is a FreeCAD term, not used in industry, and it makes people go "huh?" They understand Tools, they understand Tool Presets, and Tool Properties.

I don't understand why it was necessary to come up with `Tool controller` what is not used anywhere else
Why we can not just use `Tool` from `Tool library` without intermediaries in the form of abstractions such as `Tool controller`?

Connor9220 commented:
> > My other thought, is this: I don't think we need to use the term Tool Controller. This is a FreeCAD term, not used in industry, and it makes people go "huh?" They understand Tools, they understand Tool Presets, and Tool Properties.
> 
> I don't understand why it was necessary to come up with `Tool controller` what is not used anywhere else Why we can not just use `Tool` from `Tool library` without intermediaries in the form of abstractions such as `Tool controller`?

That is a question for @sliptonic   but, my understanding is, it allows you to adjust feeds and speeds in a single place for all operations using said tool controller. It’s a programmers paradigm born from being object oriented. It just needs to be abstracted out to make more sense to the user. I don’t think we want to mess with how it works as that touches everything including post processors.  We can just dress it up with the GUI.

sliptonic commented:
I like this improvement quite a lot.  Nice job.

Why don't I see the 'copy' button?
<img width="752" height="1114" alt="image" src="https://github.com/user-attachments/assets/1868db5a-89ca-4004-a1b1-f7290919de64" />


Without straying too far into DWG territory, I could offer some minor UI improvments.
Instead of adding a button to 'copy'  you could expand the dropdown list.
It could show for example:  


   existing TC 1
   existing TC n
   //divider
   Make a copy ....
   New TC ....


Make a copy would pop up a dialog for the user to pick one of the existing TCs,  then it would make the copy and use it.
New TC  could show the dock add a TC and use it.

davidgilkaufman commented:
<img width="954" height="810" alt="20250819_16h07m05s_grim" src="https://github.com/user-attachments/assets/58f2b450-9b11-4f5f-b048-305092b2d9b6" />

Commit pending. Also I might have figured out dynamic translations, and if I've got it right then it's not early as intimidating as I thought it was. I haven't the slightest idea how to test if I got it right though; I'm planning on just assuming.

davidgilkaufman commented:
I've implemented both "Copy" and "New" functions in the combo box, below a separator line. Copy does not invoke a new dialog; it copies the current TC. I added a label indicating how many other operations use the TC. I have not deleted the copy button (although I'm sure I should delete either the button or the combo box entry).

Let me know what you think

<img width="954" height="810" alt="20250819_16h34m16s_grim" src="https://github.com/user-attachments/assets/86d02323-fb12-4830-9544-99784a83021f" />

sliptonic commented:
> > Make a copy would pop up a dialog for the user to pick one of the existing TCs, then it would make the copy and use it.
> 
> What if it copied the currently selected tool controller? I think the main use case is to want to change feeds and speeds for the current operation without affecting other operations, notice that the current tool controller is used by other operations, and decide to "make a copy" to fix that problem. To that end, I'm unsure that jumping back into the combo box is all that intuitive, but I could build it.
> 
> Also I admit I'm not so fluent with Qt UIs as to be excited to build a dedicated dialog for this particular selection task, when there are other reasonable ways to perform the selection.

Copying the existing TC is fine.  It saves a click.  I would suggest that you make the label something like "copy of TC Default Tool..."  That's a bit more informative about what's going to happen.  

Should it ever be necessary, there is already a dialog for picking a TC from a list of TCs in the job.  It's used when you create a new operation with multiple TCs that can work.  

If the copy function is triggered from the drop down item selection, is there a reason to leave the button?  It seems to take up real estate and is unnecessary.

davidgilkaufman commented:
No, I just wanted to show both versions working. I have now removed the button, and updated text to "Copy <TC name>"

sliptonic commented:
Create a document with multiple jobs
Add operations to both.
Edit an operation in one of them. 
Use the option to add a new tool controller.

Double-clicking a tool works as expected but 'add to job' results in a prompt to pick which job to add it to.  This should be automatic.

davidgilkaufman commented:
These all sound like good ideas. I will look into making the new tool selection a blocking dialog, and capturing its output to automatically place newly created tools in the current operation's job. And I'll add ellipses in the menu.

I feel somewhat poorly about trying to implement these changes due to a growing sense of scope creep. I initially proposed a change to add feed rate overrides to operations and agreed to change the grant proposal to instead edit tool controllers from within operation UIs. Now I find myself supporting new tool creation as an additional feature in this UI and modifying tool library UI code to make this interaction a better experience, while simultaneously arguing vehemently against also rebuilding the tool bit/controller distinction as a part of the PR.

Honestly I expect all PRs to contain some level of scope creep, and probably I should calibrate my expectations to be ready for larger amounts of creep on UI work, since more people have an understanding of how the UI should work and can readily contribute valuable suggestions for improvements. But I want to put it out there that I'm having a bad time of it.

davidgilkaufman commented:
I've updated the "new tool controller..." function to display the tool picker as a dialog (no screenshots because I have a window manager that makes dialogs ugly, and that's no way to show off results). Honestly I'm shocked at how well it worked. From my past experience working with Qt UIs, I expected it to be a problem to generate blocking dialogs from a combo box change callback, and also a problem to have blocking dialogs within blocking dialogs when the user opens the tool library editor from the tool picker.

Both of these things things just worked for me, but I'd like for other people to test them. I'm running Linux and Qt6. I'd definitely like for this to be tested on Qt5. Other operating systems would be nice too, although I'm less worried about that.

> To be honest, I would support changing the 'dock' from a docked panel to a modal pop-up.

Rereading this, I now see you probably meant "for all use cases", not just "when opened from the operation UI". I've implemented the later, but it should be easy enough to always invoke my new code path and delete the dock code. My instinct is that this change does not belong in my operation/TC UI PR, but if you'd like me to add it anyway to eliminate the overhead of another PR I can do that.

davidgilkaufman commented:
I think I've fixed the white space issue. You should check too, though -- it's subtle on my screen (I think my screen is shorter, and most operations need to scroll, so there's no excess space to fill. I tested with the Engrave operation because it has few other controls to fit).

> As mentioned in the meeting, I think we should make the tool controller a first-class tab and make the edit panel visible all the time. If you want to do that in this PR, I'll hold until I hear back. Otherwise, I'll merge and we can address it in another PR later.

I'd like to merge, and address that in a later PR. Hopefully someone better than me at icon design can help with that.

> There's a subtly different dialog for picking the tool from the job setup tools page.

This dialog is new to me; I guess I have always set up my tools from the dock. My gut feeling, from having used this dialog once now, is that I prefer the version I just converted from the tool dock because it has better tool sorting.

They certainly are similar, though, and it doesn't make sense to have both of them. If it's ok with you, I'd like to merge as is, and then in a later PR perhaps we can delete the job-version dialog and replace it with the dock-version dialog. We can add a toggle option in code to not show the library editor button from the job view.

Connor9220 commented:
> I think I've fixed the white space issue. You should check too, though -- it's subtle on my screen (I think my screen is shorter, and most operations need to scroll, so there's no excess space to fill. I tested with the Engrave operation because it has few other controls to fit).
> 
> > As mentioned in the meeting, I think we should make the tool controller a first-class tab and make the edit panel visible all the time. If you want to do that in this PR, I'll hold until I hear back. Otherwise, I'll merge and we can address it in another PR later.
> 
> I'd like to merge, and address that in a later PR. Hopefully someone better than me at icon design can help with that.
> 
> > There's a subtly different dialog for picking the tool from the job setup tools page.
> 
> This dialog is new to me; I guess I have always set up my tools from the dock. My gut feeling, from having used this dialog once now, is that I prefer the version I just converted from the tool dock because it has better tool sorting.
> 
> They certainly are similar, though, and it doesn't make sense to have both of them. If it's ok with you, I'd like to merge as is, and then in a later PR perhaps we can delete the job-version dialog and replace it with the dock-version dialog. We can add a toggle option in code to not show the library editor button from the job view.

Can you post a screen shot of the dialog your currently using?

davidgilkaufman commented:
It's just the tool dock, but as a dialog

<img width="604" height="680" alt="20250825_15h47m45s_grim" src="https://github.com/user-attachments/assets/9ae84e8e-ada8-4725-b73d-20a5beab51dc" />

davidgilkaufman commented:
Is there some way for me to rerun the failed test "build_with_pixi"? It seems to have failed with a network error prior to checking out my code.

My gut feeling is there should be a button for it under the 3 dots menu to the right of the test, but I only see "view details". Perhaps I'm lacking privileges? @sliptonic

jffmichi commented:
I meant to say "some people". I know because I'm one of them. I like that I can open the dock, add all tools with a simple double click while scrolling through the list once and then close it again.

sliptonic commented:
> I meant to say "some people". I know because I'm one of them. I like that I can open the dock, add all tools with a simple double click while scrolling through the list once and then close it again.

Cool.  Yes, that's a valid use-case and I think it can be achieved without docking the widget.  Ctrl+select all the tools you want and then approve the selection.  I think the dialog should work the same whether you come in from the job edit panel or the operation panel.

jffmichi commented:
But why change a perfectly working thing that "some people" are already used to? The dialog and dock can coexist without much additional code. "Other people" can just use the dialog and never have to see the dock if they don't want to.

sliptonic commented:
> But why change a perfectly working thing that "some people" are already used to? The dialog and dock can coexist without much additional code. "Other people" can just use the dialog and never have to see the dock if they don't want to.

Two reasons:

"much additional code" is still additional code that has to be maintained.  

More importantly, software developers think we're doing a favor to users by allowing multiple ways to do the same thing.  We're not.  Users have to figure out why there are two ways and how they differ.  This is cognitive load and it's annoying. Users will adapt to the UI if it is efficient and intuitive. 

If it were me, I would only allow a single place to do this, probably in the job setup dialog.  But I recognize that there are two natural workflows that people follow and both are intuitive.
- Some people set up all the tools at the beginning and then start building ops.
- Some people start building ops and add the tools as needed.  

Neither is better than the other so we have to allow for both.  If the two UI panels are nearly identical, there's no additional cognitive burden.

After this, there will be three ways.  dock, job setup, op.   The job dialog is different. 
The dock and modal dialog do the same thing but now we have to carry the logic for what to do if the dock is already visible.  It's just annoying GUI state logic that is buggy and adds very little value.

If you keep the dock, then you need a gui element to open/close it.  That's the existing toolbar button.  
The only way to get to the library editor is by opening the dock, then opening the editor (two clicks)

If we eliminate the dock, we eliminate the toolbar button or could repurpose it to open the library editor directly.  

BTW,  I can always be convinced otherwise on matters of GUI and will defer to the @FreeCAD/design-working-group if they have opinions.

davidgilkaufman commented:
> Quick little patch file to fix the dialog padding and size issues.. however, I detest hard coding this stuff.. any better options?

Thanks Billy; patch accepted.

I'm not a UI person, and also I run a tiling window manager, so unfortunately both of these things make me poor candidates for these questions (the dialog way _always_ large and readable for me!).

I'll take a look at your other issue now

davidgilkaufman commented:
> When I edit / Copy the tool controller, it looses the original Spindle Speed and Vertical feed.

Good catch. I fixed this bug. I think it was too introduced due to a subtle ui interaction when I switched from the copy button to the copy combo item.



> Hitting the ESC key while in the Expression Editor kicks you out of the operation completely.

This one I'm throwing in the towel after debugging it for an hour and a half. It's a pre-existing bug, not related to my change. You can also observe this behavior other UI elements, even outside of CAM (i.e. Pad -> edit formula length -> hit escape). This bug does not exist on 1.0.1, but it's not new in my change either. I have reported it upstream #23518. My diagnosis is that the escape handler that covers closing these panels is overzealous, but I don't understand freecad's chain of UI elements well enough to know how to resolve it.
